### ISSUE TITLE
Create builder failed with 401 (Unauthorized) when using AKS MI

### ISSUE THREAD

**Issue by frankliu20** [pivotal/kpack#907]
We are using [AKS MI](https://docs.microsoft.com/en-us/azure/aks/use-managed-identity) to do ACR auth when pulling image.  The kpack-controller works well with AKS MI when it just stated, and it can create kpack-builder successfully. **However, after some minutes (around 1h)**, when we try to create a new builder, the kpack-controller failed with some 401 (Unauthorized) error.
Check this log
```
{"severity":"ERROR","timestamp":"2021-12-13T06:15:18.089019623Z","logger":"controller","caller":"controller/controller.go:548","message":"Reconcile error","commit":"121ba5c","duration":"4.011929985s","error":"Get \"https://xxxacr.azurecr.io/v2/kpack/lifecycle/blobs/sha256:cfc3c2aa5deeb84c2498296c88c64cc20089461e9573849e84dc1f12b4fc7857\": POST https://xxxacr.azurecr.io/oauth2/token: unexpected status code 401 Unauthorized","stacktrace":"knative.dev/pkg/controller.(*Impl).handleErr\n\t/home/cnb/go/pkg/mod/knative.dev/pkg@v0.0.0-20210331065221-952fdd90dbb0/controller/controller.go:548\nknative.dev/pkg/controller.(*Impl).processNextWorkItem\n\t/home/cnb/go/pkg/mod/knative.dev/pkg@v0.0.0-20210331065221-952fdd90dbb0/controller/controller.go:531\nknative.dev/pkg/controller.(*Impl).RunContext.func3\n\t/home/cnb/go/pkg/mod/knative.dev/pkg@v0.0.0-20210331065221-952fdd90dbb0/controller/controller.go:468"}
```
It seems the kpack-controller **does not** try to acquire a new token when the initial acr token is expired. Or something goes wrong when trying to get the new token.

We are using kpack version `0.3.1`, and the kpack-controller looks like this:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: kpack-controller
  namespace: kpack
spec:
  replicas: 1
  selector:
    matchLabels:
      app: kpack-controller
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: kpack-controller
        version: 0.3.1
    spec:
      containers:
      - env:
        - name: AZURE_CONTAINER_REGISTRY_CONFIG
          value: /etc/kubernetes/azure.json
        - name: CONFIG_LOGGING_NAME
          value: config-logging
        - name: CONFIG_OBSERVABILITY_NAME
          value: config-observability
        - name: METRICS_DOMAIN
          value: kpack.io
        - name: SYSTEM_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        - name: BUILD_INIT_IMAGE
          valueFrom:
            configMapKeyRef:
              key: image
              name: build-init-image
        - name: REBASE_IMAGE
          valueFrom:
            configMapKeyRef:
              key: image
              name: rebase-image
        - name: COMPLETION_IMAGE
          valueFrom:
            configMapKeyRef:
              key: image
              name: completion-image
        - name: LIFECYCLE_IMAGE
          valueFrom:
            configMapKeyRef:
              key: image
              name: lifecycle-image
        image: xxxacr.azurecr.io/kpack/controller:2021-0709-0627
        imagePullPolicy: IfNotPresent
        name: controller
        resources:
          limits:
            cpu: 100m
            memory: 500Mi
          requests:
            cpu: 20m
            memory: 100Mi
        securityContext:
          runAsGroup: 0
          runAsUser: 0
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/kubernetes/azure.json
          name: azure-config-file
          readOnly: true
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      serviceAccount: controller
      serviceAccountName: controller
      terminationGracePeriodSeconds: 30
      volumes:
      - hostPath:
          path: /etc/kubernetes/azure.json
          type: File
        name: azure-config-file

```

Here is the mounted `azure.json` file 
```
{
    "cloud": "AzurePublicCloud",
    "tenantId": "xxxtenantid",
    "subscriptionId": "xxxsubid",
    "aadClientId": "msi",
    "aadClientSecret": "msi",
    "resourceGroup": "aks_nodes_xxx",
    "location": "eastus",
    "vmType": "vmss",
    "subnetName": "xxx",
    "securityGroupName": "xxx-nsg",
    "vnetName": "xxx-vnet",
    "vnetResourceGroup": "xxxgroup",
    "routeTableName": "xxxx-routetable",
    "primaryAvailabilitySetName": "",
    "primaryScaleSetName": "aks-default-xxx-vmss",
    "cloudProviderBackoffMode": "v2",
    "cloudProviderBackoff": true,
    "cloudProviderBackoffRetries": 6,
    "cloudProviderBackoffDuration": 5,
    "cloudProviderRateLimit": true,
    "cloudProviderRateLimitQPS": 30,
    "cloudProviderRateLimitBucket": 300,
    "cloudProviderRateLimitQPSWrite": 30,
    "cloudProviderRateLimitBucketWrite": 300,
    "useManagedIdentityExtension": true,
    "userAssignedIdentityID": "2594db5f-7b9e-4319-950d-808aef394f28",
    "useInstanceMetadata": true,
    "loadBalancerSku": "Standard",
    "disableOutboundSNAT": false,
    "excludeMasterFromStandardLB": true,
    "providerVaultName": "",
    "maximumLoadBalancerRuleCount": 250,
    "providerKeyName": "k8s",
    "providerKeyVersion": ""
}

```

Could you pls help check and give some help on this?

**Comment by tylerphelan** [comment]
@frankliu20 this auth error happens intermittently? So other builders can be created? You can also post in the buildpacks.slack.com #kpack channel for faster support.

**Comment by allxiao** [comment]
It seems this is caused by kpack not refreshing the token.

When kpack is started / restarted, it works well for a while. After some time (typically around 1 hour), all builder creation results in 401 error from the private Azure Container Registry (ACR).

The setting `useManagedIdentityExtension: true` is present in the configuration file, and kpack does working for a short time after start, so it's able to utilize the managed identity to authenticate to ACR. However, it seems the token is cached and not refreshed, so later when the token expires, we are not able to connect to the ACR to pull images which are required for the builder creation.

**Comment by allxiao** [comment]
Regarding the slack channel, it seems we do not have access to the buildpacks workspace, so we are not able to see / post anything there.

**Comment by FrankLiu4138** [comment]
> @frankliu20 this auth error happens intermittently? So other builders can be created? You can also post in the buildpacks.slack.com #kpack channel for faster support.

Not intermittently, when this happens we can not create new builders due to the auth failure. The only workaround is restart the koack controller.

**Comment by frankliu20** [comment]
> @frankliu20 this auth error happens intermittently? So other builders can be created? You can also post in the buildpacks.slack.com #kpack channel for faster support.

Not intermittently, when this happens we can **not** create new builders due to the auth failure. The only workaround is restart the kpack controller.

**Comment by tylerphelan** [comment]
@allxiao @frankliu20 

> Regarding the slack channel, it seems we do not have access to the buildpacks workspace, so we are not able to see / post anything there.

the slack domain is buildpacks.slack.com which is public. The channel is #kpack, if you give me your slack handle in buildpacks.slack.com I can add you to the channel

**Comment by allxiao** [comment]
Hi @matthewmcnew , could you please help to reproduce this issue and provide more context about the root cause? kpack is like a black box for us. We are willing to reach to the upstream for the resolution, but if we do not have any details but only the surface problem, it would be difficult to move forward in the conversation with them.

We can prepare and share the AKS environment in which the kpack-controller is not working, so that you can test in this cluster, for example, using a private built kpack with verbose logging.

Thanks,

**Comment by allxiao** [comment]
The klog verbosity setting is not exposed to kpack-controller arguments ([`--v`](https://github.com/kubernetes/klog/blob/main/klog.go#L1408)) so we are not able to turn on verbose logging for kpack-controller to get more detailed information.

**Comment by tylerphelan** [comment]
This has been resolved in kpack [0.5.2](https://github.com/pivotal/kpack/releases/tag/v0.5.2)

Let us know if you are still seeing this issue!
